include/libfreenect2/libfreenect2.hpp

```diff
@@ -274,6 +278,49 @@ class LIBFREENECT2_API Freenect2
   Freenect2& operator=(const Freenect2&);
 };
 
+class LIBFREENECT2_API Freenect2ReplayDevice: public Freenect2Device
+{
+public:
+  Freenect2ReplayDevice(char* directory, PacketPipeline* pipeline);
+ 
+  virtual std::string getSerialNumber();
+  virtual std::string getFirmwareVersion();
+
+  virtual ColorCameraParams getColorCameraParams();
+  virtual IrCameraParams getIrCameraParams();
+
+  virtual void setColorCameraParams(const ColorCameraParams &params);
+  virtual void setIrCameraParams(const IrCameraParams &params);
+
+  void setColorFrameListener(FrameListener* listener);
+  void setIrAndDepthFrameListener(FrameListener* listener);
+
+  virtual void setConfiguration(const Config &config);
+  
+  void loadP0Tables(unsigned char* buffer, size_t buffer_length);
+  void loadXZTables(const float *xtable, const float *ztable);
+  void loadLookupTable(const short *lut);
+
+  bool processRawFrame(Frame::Type type, Frame* frame);  
+
+  virtual bool start();
+  virtual bool startStreams(bool rgb, bool depth);
+  virtual bool stop();
+  virtual bool run();
+
+protected:
+  void processRgbFrame(Frame* frame);
+  void processDepthFrame(Frame* frame);
+
+private:
+  PacketPipeline* pipeline_;
+  char* directory_;
+  tthread::thread* t_;
```

Comment:
@xlz this (`Freenect2Replay`) is a major one. I would propose instead something akin to a factory method or even a factory class to create desired device types. My preferred approach is to teach `Freenect2::openDevice()` and friends to open any kind of device specified by its enumeration type or whatever, defaulting to the regular device. This way, e.g., using Protonect, it would be a drop-in replacement to select a Replay device as opposed to a regular device and treat them uniformly.

> Also move the declaration of Freenect2ReplayDevice out of the header to libfreenect2.cpp after Freenect2DeviceImpl

Will do.

Reply:
I thought of this idea of overloading Freenect2::openDevice(). Freenect2's constructor initializes libusb context and start a USb thread. For replay usage that is unnecessary.

Reply:
`Freenect2::Freenect2(NULL)` would indicate your USB context is not there and you initialize the Replay device instead?

Reply:
Freenect2::Freenect2(NULL) means "managed USB context" which automatically creates it. Non-NULL is using an external USB context. You can picture `Freenect2` as `Freenect2USB` in your mind but we won't rename it though.

Reply:
So then we need to adjust the design with a parent abstract interface for `Freenect2` (aka Freenect2USB) and Freenect2Replay and use that, I guess.

Reply:
TODO

Reply:
For the sake of ABI stability (generally you want to just add stuff not change any stuff at interface level) just ignore the aesthetics. We discussed adding Freenect2::openReplayDevice() involves unnecessary work, and we don't want to introduce additional change just to fix that unnecessary work.

A new Freenect2Replay class just works. I don't see any problem with it.

Reply:
@xlz -- implemented `Freenect2Replay` extrapolating from your proposed API design below. I consider this thread item is done; please comment on the updated commit below.
DONE